File: Sources/ReminderAI/ExerciseInterpreter.swift
import Foundation
import ReminderCore

/// Turns free text into exercises using something cleverer than the built-in
/// parser.
///
/// A protocol so the import sheet can be driven by a stub in tests, previews
/// and snapshots without a network or an API key — the same seam
/// `SpeechCoaching` gives the voice coach.
public protocol ExerciseInterpreting: Sendable {
    /// Interprets `text`, or throws if it cannot.
    ///
    /// Implementations return rows already passed through
    /// `Exercise.normalized(_:)`, so a caller can never receive an exercise the
    /// editor would reject however the model answered.
    func interpret(_ text: String) async throws -> [Exercise]
}

/// The models offered for interpreting exercise text.
///
/// Deliberately a short list rather than a free-text field: a typo in a model
/// name surfaces as an opaque API error, and none of the larger models are
/// worth their cost for pulling six numbers out of a paragraph.
public enum AIImportModel: String, CaseIterable, Identifiable, Sendable {
    case nano = "gpt-5-nano"
    case luna = "gpt-5.6-luna"
    case mini = "gpt-5-mini"

    public var id: String { rawValue }

    /// The one used when the user has not chosen.
    ///
    /// Not the cheapest on the list, deliberately. Measured on a three-exercise
    /// paragraph, `gpt-5-nano` took 31 s and `gpt-5.6-luna` 6 s: the nano tier
    /// spends far more reasoning effort on a small extraction than the task
    /// warrants, and 30 s of staring at a spinner is the difference between a
    /// feature that saves time and one that does not. Luna is still well under
    /// a tenth of a cent per import, and read the same text more accurately.
    public static let `default`: AIImportModel = .luna

    /// Resolves the stored setting, falling back to the default for `nil` or
    /// for a model that has since been retired.
    public static func resolve(_ identifier: String?) -> AIImportModel {
        guard let identifier, let model = AIImportModel(rawValue: identifier) else {
            return .default
        }
        return model
    }

    public var title: String {
        switch self {
        case .nano: return "GPT-5 nano (cheapest, but slow)"
        case .luna: return "GPT-5.6 Luna (recommended)"
        case .mini: return "GPT-5 mini"
        }
    }

    public var detail: String {
        switch self {
        case .nano: return "Cheapest, but often takes 30 seconds or more to answer."
        case .luna: return "Reads flowing prose well and answers in a few seconds."
        case .mini: return "Better with unusual wording or long programmes."
        }
    }
}

/// Something went wrong talking to OpenAI, phrased for the person who pasted
/// the text rather than for a log.
public enum AIImportError: LocalizedError, Equatable {
    case missingKey
    case unauthorized
    case rateLimited
    case offline
    case timedOut
    case server(status: Int, message: String?)
    case unreadableResponse
    case nothingFound

    public var errorDescription: String? {
        switch self {
        case .missingKey:
            return "Add an OpenAI API key in Settings to interpret text with AI."
        case .unauthorized:
            return "That API key was rejected. Check it in Settings."
        case .rateLimited:
            return "OpenAI is rate limiting this key. Try again in a moment."
        case .offline:
            return "No connection to OpenAI. The built-in parser still works."
        case .timedOut:
            return "OpenAI took too long to answer. Try again, or use Read Text."
        case let .server(status, message):
            return message ?? "OpenAI returned an error (\(status))."
        case .unreadableResponse:
            return "OpenAI's reply could not be read."
        case .nothingFound:
            return "No exercises were found in that text."
        }
    }
}

/// Interprets exercise text with OpenAI's Responses API, using Structured
/// Outputs so the reply is guaranteed to match the schema below rather than
/// being prose we then have to salvage.
public struct OpenAIExerciseInterpreter: ExerciseInterpreting {
    private let apiKey: String
    private let model: AIImportModel
    private let session: URLSession
    private let endpoint: URL

    public init(
        apiKey: String,
        model: AIImportModel = .default,
        session: URLSession = .shared,
        endpoint: URL = URL(string: "https://api.openai.com/v1/responses")!
    ) {
        self.apiKey = apiKey
        self.model = model
        self.session = session
        self.endpoint = endpoint
    }

    public func interpret(_ text: String) async throws -> [Exercise] {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { throw AIImportError.nothingFound }
        guard !apiKey.isEmpty else { throw AIImportError.missingKey }

        var request = URLRequest(url: endpoint)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.httpBody = try JSONSerialization.data(withJSONObject: body(for: trimmed))
        // Generous: a reasoning model working through a whole programme
        // regularly takes over 30 seconds, and a slow answer is still the
        // answer the person asked for.
        request.timeoutInterval = 120

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: request)
        } catch let error as URLError where Self.offlineCodes.contains(error.code) {
            throw AIImportError.offline
        } catch let error as URLError where error.code == .timedOut {
            throw AIImportError.timedOut
        }

        guard let http = response as? HTTPURLResponse else { throw AIImportError.unreadableResponse }
        switch http.statusCode {
        case 200...299: break
        case 401, 403: throw AIImportError.unauthorized
        case 429: throw AIImportError.rateLimited
        default: throw AIImportError.server(status: http.statusCode, message: Self.message(in: data))
        }

        let payload = try Self.structuredText(in: data)
        let decoded = try Self.decode(payload)
        guard let normalized = Exercise.normalized(decoded), !normalized.isEmpty else {
            throw AIImportError.nothingFound
        }
        return normalized
    }

    /// Codes that really do mean "the request never got there". `.timedOut`
    /// is deliberately absent: it means OpenAI was reachable but slow, and
    /// reporting that as a connection fault sends people to check a network
    /// that was never broken.
    private static let offlineCodes: Set<URLError.Code> = [
        .notConnectedToInternet, .networkConnectionLost, .cannotFindHost,
        .cannotConnectToHost, .dataNotAllowed, .internationalRoamingOff,
    ]

    // MARK: - Request

    private func body(for text: String) -> [String: Any] {
        [
            "model": model.rawValue,
            "input": [
                ["role": "system", "content": Self.systemPrompt],
                ["role": "user", "content": text],
            ],
            "text": [
                "format": [
                    "type": "json_schema",
                    "name": "exercises",
                    "strict": true,
                    "schema": Self.schema,
                ]
            ],
        ]
    }

    /// Extraction, not authoring. The model must not invent a hold time that
    /// was never prescribed — a wrong number here becomes a wrong instruction
    /// during someone's rehabilitation.
    static let systemPrompt = """
        You convert physiotherapy exercise instructions into structured data.

        Rules:
        - Extract only what the text states. Never invent or infer timings.
        - When a value is not stated use these defaults: sets 3, reps 10, \
        holdSeconds 0, restBetweenRepsSeconds 0, restBetweenSetsSeconds 0.
        - All durations are in whole seconds. Convert minutes.
        - "name" is a short exercise name, without counts or timings.
        - "instructions" is any remaining guidance on how to perform it, \
        verbatim where possible, or an empty string.
        - One object per distinct exercise, in the order given.
        - If the text describes no exercises, return an empty array.
        """

    /// Mirrors `Exercise`'s stored fields. `strict: true` requires every
    /// property to be listed as required and additional ones forbidden, so the
    /// reply always decodes.
    static let schema: [String: Any] = [
        "type": "object",
        "additionalProperties": false,
        "required": ["exercises"],
        "properties": [
            "exercises": [
                "type": "array",
                "items": [
                    "type": "object",
                    "additionalProperties": false,
                    "required": [
                        "name", "instructions", "sets", "reps",
                        "holdSeconds", "restBetweenRepsSeconds", "restBetweenSetsSeconds",
                    ],
                    "properties": [
                        "name": ["type": "string"],
                        "instructions": ["type": "string"],
                        "sets": ["type": "integer"],
                        "reps": ["type": "integer"],
                        "holdSeconds": ["type": "integer"],
                        "restBetweenRepsSeconds": ["type": "integer"],
                        "restBetweenSetsSeconds": ["type": "integer"],
                    ],
                ],
            ]
        ],
    ]

    // MARK: - Response

    /// Digs the JSON string out of the Responses envelope.
    ///
    /// `output_text` is the convenience field; the walk over `output` is the
    /// documented shape and the fallback when it is absent.
    static func structuredText(in data: Data) throws -> Data {
        guard let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw AIImportError.unreadableResponse
        }
        if let text = root["output_text"] as? String, !text.isEmpty {
            return Data(text.utf8)
        }
        guard let output = root["output"] as? [[String: Any]] else {
            throw AIImportError.unreadableResponse
        }
        for item in output {
            guard let contents = item["content"] as? [[String: Any]] else { continue }
            for content in contents {
                if let text = content["text"] as? String, !text.isEmpty {
                    return Data(text.utf8)
                }
            }
        }
        throw AIImportError.unreadableResponse
    }

    /// Decodes the model's payload into exercises.
    ///
    /// Decoded field by field rather than through `Exercise`'s own `Codable`,
    /// because the model does not send an `id` and must never be able to set
    /// one — identity is ours to assign.
    static func decode(_ payload: Data) throws -> [Exercise] {
        guard let root = try? JSONSerialization.jsonObject(with: payload) as? [String: Any],
              let rows = root["exercises"] as? [[String: Any]] else {
            throw AIImportError.unreadableResponse
        }
        return rows.compactMap { row in
            guard let name = row["name"] as? String else { return nil }
            return Exercise(
                name: name,
                instructions: row["instructions"] as? String ?? "",
                sets: integer(row["sets"]) ?? 3,
                reps: integer(row["reps"]) ?? 10,
                holdSeconds: integer(row["holdSeconds"]) ?? 0,
                restBetweenRepsSeconds: integer(row["restBetweenRepsSeconds"]) ?? 0,
                restBetweenSetsSeconds: integer(row["restBetweenSetsSeconds"]) ?? 0
            )
        }
    }

    /// Accepts a number written as a JSON number or as a string — models are
    /// not always consistent, and a quoted "3" should not lose the count.
    private static func integer(_ value: Any?) -> Int? {
        if let number = value as? Int { return number }
        if let number = value as? Double { return Int(number) }
        if let text = value as? String { return Int(text) }
        return nil
    }

    /// The human-readable half of an OpenAI error body, when there is one.
    static func message(in data: Data) -> String? {
        guard let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let error = root["error"] as? [String: Any],
              let message = error["message"] as? String else { return nil }
        return message
    }
}

/// A stand-in interpreter for previews, snapshots and tests: answers instantly
/// from the built-in parser instead of reaching the network, so nothing that
/// renders the UI can make a request or need a key.
public struct PreviewExerciseInterpreter: ExerciseInterpreting {
    private let result: [Exercise]?

    /// - Parameter result: what to return; `nil` parses the text locally.
    public init(result: [Exercise]? = nil) { self.result = result }

    public func interpret(_ text: String) async throws -> [Exercise] {
        if let result { return result }
        return ExerciseImporter.parse(text)
    }
}

File: Windows/Pauselet.Core/ExerciseInterpreter.cs
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;

namespace Pauselet.Core;

/// <summary>
/// Turns free text into exercises using something cleverer than the built-in
/// parser. (Mirrors ExerciseInterpreting in ExerciseInterpreter.swift.)
///
/// An interface so the import dialog can be driven by a stub without a network
/// or an API key — the same seam <c>ISpeechCoaching</c> gives the voice coach.
/// </summary>
public interface IExerciseInterpreter
{
    /// <summary>
    /// Interprets <paramref name="text"/>, or throws
    /// <see cref="AIImportException"/> if it cannot.
    ///
    /// Implementations return rows already passed through
    /// <see cref="Exercise.Normalized"/>, so a caller can never receive an
    /// exercise the editor would reject however the model answered.
    /// </summary>
    Task<IReadOnlyList<Exercise>> InterpretAsync(
        string text, CancellationToken cancellationToken = default);
}

/// <summary>
/// The models offered for interpreting exercise text.
///
/// Deliberately a short list rather than a free-text field: a typo in a model
/// name surfaces as an opaque API error, and none of the larger models are
/// worth their cost for pulling six numbers out of a paragraph.
/// </summary>
public sealed record AIImportModel(string Id, string Title, string Detail)
{
    public static readonly AIImportModel Nano = new(
        "gpt-5-nano",
        "GPT-5 nano (cheapest, but slow)",
        "Cheapest, but often takes 30 seconds or more to answer.");

    public static readonly AIImportModel Luna = new(
        "gpt-5.6-luna",
        "GPT-5.6 Luna (recommended)",
        "Reads flowing prose well and answers in a few seconds.");

    public static readonly AIImportModel Mini = new(
        "gpt-5-mini",
        "GPT-5 mini",
        "Better with unusual wording or long programmes.");

    /// <summary>In the order the picker shows them, matching the Mac.</summary>
    public static readonly IReadOnlyList<AIImportModel> All = [Nano, Luna, Mini];

    /// <summary>
    /// The one used when the user has not chosen.
    ///
    /// Not the cheapest on the list, deliberately. Measured on a three-exercise
    /// paragraph, gpt-5-nano took 31 s and gpt-5.6-luna 6 s: the nano tier
    /// spends far more reasoning effort on a small extraction than the task
    /// warrants, and 30 s of staring at a spinner is the difference between a
    /// feature that saves time and one that does not.
    /// </summary>
    public static AIImportModel Default => Luna;

    /// <summary>
    /// Resolves the stored setting, falling back to the default for
    /// <c>null</c> or for a model that has since been retired.
    /// </summary>
    public static AIImportModel Resolve(string? identifier) =>
        All.FirstOrDefault(model => model.Id == identifier) ?? Default;
}

/// <summary>
/// Something went wrong talking to OpenAI, phrased for the person who pasted
/// the text rather than for a log.
/// </summary>
public sealed class AIImportException : Exception
{
    public enum Reason
    {
        MissingKey,
        Unauthorized,
        RateLimited,
        /// <summary>The request never got there.</summary>
        Offline,
        /// <summary>OpenAI was reachable but slow — a different fault from offline.</summary>
        TimedOut,
        Server,
        UnreadableResponse,
        NothingFound,
    }

    public Reason FailureReason { get; }

    private AIImportException(Reason reason, string message) : base(message) =>
        FailureReason = reason;

    public static AIImportException MissingKey() => new(
        Reason.MissingKey, "Add an OpenAI API key in Settings to interpret text with AI.");

    public static AIImportException Unauthorized() => new(
        Reason.Unauthorized, "That API key was rejected. Check it in Settings.");

    public static AIImportException RateLimited() => new(
        Reason.RateLimited, "OpenAI is rate limiting this key. Try again in a moment.");

    public static AIImportException Offline() => new(
        Reason.Offline, "No connection to OpenAI. The built-in parser still works.");

    public static AIImportException TimedOut() => new(
        Reason.TimedOut, "OpenAI took too long to answer. Try again, or use Read Text.");

    public static AIImportException Server(int status, string? message) => new(
        Reason.Server, message ?? $"OpenAI returned an error ({status}).");

    public static AIImportException UnreadableResponse() => new(
        Reason.UnreadableResponse, "OpenAI's reply could not be read.");

    public static AIImportException NothingFound() => new(
        Reason.NothingFound, "No exercises were found in that text.");
}

/// <summary>
/// Interprets exercise text with OpenAI's Responses API, using Structured
/// Outputs so the reply is guaranteed to match the schema below rather than
/// being prose we then have to salvage.
/// </summary>
public sealed class OpenAIExerciseInterpreter : IExerciseInterpreter
{
    /// <summary>
    /// Generous: a reasoning model working through a whole programme regularly
    /// takes over 30 seconds, and a slow answer is still the answer the person
    /// asked for.
    /// </summary>
    public static readonly TimeSpan Timeout = TimeSpan.FromSeconds(120);

    /// <summary>
    /// One client for the app's lifetime. A new <c>HttpClient</c> per request
    /// holds its socket open past disposal and eventually exhausts the port
    /// pool; one shared instance is the documented way to use it, and there is
    /// only ever one endpoint here.
    /// </summary>
    private static readonly HttpClient Shared = new() { Timeout = Timeout };

    private readonly string _apiKey;
    private readonly AIImportModel _model;
    private readonly HttpClient _client;
    private readonly Uri _endpoint;

    public OpenAIExerciseInterpreter(
        string apiKey,
        AIImportModel? model = null,
        HttpClient? client = null,
        Uri? endpoint = null)
    {
        _apiKey = apiKey;
        _model = model ?? AIImportModel.Default;
        // Tests inject their own; everything else shares the one above.
        _client = client ?? Shared;
        _endpoint = endpoint ?? new Uri("https://api.openai.com/v1/responses");
    }

    public async Task<IReadOnlyList<Exercise>> InterpretAsync(
        string text, CancellationToken cancellationToken = default)
    {
        var trimmed = text.Trim();
        if (trimmed.Length == 0) throw AIImportException.NothingFound();
        if (_apiKey.Length == 0) throw AIImportException.MissingKey();

        using var request = new HttpRequestMessage(HttpMethod.Post, _endpoint)
        {
            Content = new StringContent(Body(trimmed), Encoding.UTF8, "application/json"),
        };
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _apiKey);

        HttpResponseMessage response;
        try
        {
            response = await _client.SendAsync(request, cancellationToken).ConfigureAwait(false);
        }
        catch (TaskCanceledException) when (!cancellationToken.IsCancellationRequested)
        {
            // HttpClient reports its own timeout as a cancellation; a genuine
            // caller cancellation is the guard above. Reported as timed out
            // rather than offline: saying "no connection" about a reachable
            // but slow service sends people to debug a working network.
            throw AIImportException.TimedOut();
        }
        catch (HttpRequestException)
        {
            throw AIImportException.Offline();
        }

        using (response)
        {
            var data = await response.Content
                .ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
            if (!response.IsSuccessStatusCode)
            {
                throw response.StatusCode switch
                {
                    HttpStatusCode.Unauthorized or HttpStatusCode.Forbidden =>
                        AIImportException.Unauthorized(),
                    HttpStatusCode.TooManyRequests => AIImportException.RateLimited(),
                    _ => AIImportException.Server((int)response.StatusCode, MessageIn(data)),
                };
            }

            var payload = StructuredText(data);
            var decoded = Decode(payload);
            var normalized = Exercise.Normalized(decoded);
            if (normalized is not { Count: > 0 }) throw AIImportException.NothingFound();
            return normalized;
        }
    }

    // MARK: - Request

    private string Body(string text)
    {
        // Written with the writer rather than an anonymous object so the schema
        // below reads as the JSON OpenAI documents, not as a C# shape that
        // happens to serialize to it.
        var buffer = new MemoryStream();
        using (var writer = new Utf8JsonWriter(buffer))
        {
            writer.WriteStartObject();
            writer.WriteString("model", _model.Id);

            writer.WriteStartArray("input");
            writer.WriteStartObject();
            writer.WriteString("role", "system");
            writer.WriteString("content", SystemPrompt);
            writer.WriteEndObject();
            writer.WriteStartObject();
            writer.WriteString("role", "user");
            writer.WriteString("content", text);
            writer.WriteEndObject();
            writer.WriteEndArray();

            writer.WriteStartObject("text");
            writer.WriteStartObject("format");
            writer.WriteString("type", "json_schema");
            writer.WriteString("name", "exercises");
            writer.WriteBoolean("strict", true);
            writer.WritePropertyName("schema");
            writer.WriteRawValue(Schema);
            writer.WriteEndObject();
            writer.WriteEndObject();

            writer.WriteEndObject();
        }
        return Encoding.UTF8.GetString(buffer.ToArray());
    }

    /// <summary>
    /// Extraction, not authoring. The model must not invent a hold time that
    /// was never prescribed — a wrong number here becomes a wrong instruction
    /// during someone's rehabilitation.
    /// </summary>
    internal const string SystemPrompt = """
        You convert physiotherapy exercise instructions into structured data.

        Rules:
        - Extract only what the text states. Never invent or infer timings.
        - When a value is not stated use these defaults: sets 3, reps 10, holdSeconds 0, restBetweenRepsSeconds 0, restBetweenSetsSeconds 0.
        - All durations are in whole seconds. Convert minutes.
        - "name" is a short exercise name, without counts or timings.
        - "instructions" is any remaining guidance on how to perform it, verbatim where possible, or an empty string.
        - One object per distinct exercise, in the order given.
        - If the text describes no exercises, return an empty array.
        """;

    /// <summary>
    /// Mirrors <see cref="Exercise"/>'s stored fields. <c>strict: true</c>
    /// requires every property to be listed as required and additional ones
    /// forbidden, so the reply always decodes.
    /// </summary>
    internal const string Schema = """
        {
          "type": "object",
          "additionalProperties": false,
          "required": ["exercises"],
          "properties": {
            "exercises": {
              "type": "array",
              "items": {
                "type": "object",
                "additionalProperties": false,
                "required": [
                  "name", "instructions", "sets", "reps",
                  "holdSeconds", "restBetweenRepsSeconds", "restBetweenSetsSeconds"
                ],
                "properties": {
                  "name": { "type": "string" },
                  "instructions": { "type": "string" },
                  "sets": { "type": "integer" },
                  "reps": { "type": "integer" },
                  "holdSeconds": { "type": "integer" },
                  "restBetweenRepsSeconds": { "type": "integer" },
                  "restBetweenSetsSeconds": { "type": "integer" }
                }
              }
            }
          }
        }
        """;

    // MARK: - Response

    /// <summary>
    /// Digs the JSON string out of the Responses envelope.
    ///
    /// <c>output_text</c> is the convenience field; the walk over
    /// <c>output</c> is the documented shape and the fallback when it is absent.
    /// </summary>
    internal static string StructuredText(string data)
    {
        JsonDocument document;
        try
        {
            document = JsonDocument.Parse(data);
        }
        catch (JsonException)
        {
            throw AIImportException.UnreadableResponse();
        }
        using (document)
        {
            var root = document.RootElement;
            if (root.ValueKind != JsonValueKind.Object) throw AIImportException.UnreadableResponse();

            if (root.TryGetProperty("output_text", out var outputText)
                && outputText.ValueKind == JsonValueKind.String
                && outputText.GetString() is { Length: > 0 } text)
            {
                return text;
            }

            if (!root.TryGetProperty("output", out var output)
                || output.ValueKind != JsonValueKind.Array)
            {
                throw AIImportException.UnreadableResponse();
            }
            foreach (var item in output.EnumerateArray())
            {
                if (!item.TryGetProperty("content", out var contents)
                    || contents.ValueKind != JsonValueKind.Array)
                {
                    continue;
                }
                foreach (var content in contents.EnumerateArray())
                {
                    if (content.TryGetProperty("text", out var value)
                        && value.ValueKind == JsonValueKind.String
                        && value.GetString() is { Length: > 0 } found)
                    {
                        return found;
                    }
                }
            }
            throw AIImportException.UnreadableResponse();
        }
    }

    /// <summary>
    /// Decodes the model's payload into exercises.
    ///
    /// Decoded field by field rather than through the data file's own decoder,
    /// because the model does not send an id and must never be able to set one
    /// — identity is ours to assign.
    /// </summary>
    internal static IReadOnlyList<Exercise> Decode(string payload)
    {
        JsonDocument document;
        try
        {
            document = JsonDocument.Parse(payload);
        }
        catch (JsonException)
        {
            throw AIImportException.UnreadableResponse();
        }
        using (document)
        {
            var root = document.RootElement;
            if (root.ValueKind != JsonValueKind.Object
                || !root.TryGetProperty("exercises", out var rows)
                || rows.ValueKind != JsonValueKind.Array)
            {
                throw AIImportException.UnreadableResponse();
            }

            var exercises = new List<Exercise>();
            foreach (var row in rows.EnumerateArray())
            {
                if (row.ValueKind != JsonValueKind.Object) continue;
                if (!row.TryGetProperty("name", out var name)
                    || name.ValueKind != JsonValueKind.String)
                {
                    continue;
                }
                exercises.Add(new Exercise
                {
                    Name = name.GetString() ?? "",
                    Instructions = Text(row, "instructions") ?? "",
                    Sets = Integer(row, "sets") ?? 3,
                    Reps = Integer(row, "reps") ?? 10,
                    HoldSeconds = Integer(row, "holdSeconds") ?? 0,
                    RestBetweenRepsSeconds = Integer(row, "restBetweenRepsSeconds") ?? 0,
                    RestBetweenSetsSeconds = Integer(row, "restBetweenSetsSeconds") ?? 0,
                });
            }
            return exercises;
        }
    }

    private static string? Text(JsonElement row, string name) =>
        row.TryGetProperty(name, out var value) && value.ValueKind == JsonValueKind.String
            ? value.GetString()
            : null;

    /// <summary>
    /// Accepts a number written as a JSON number or as a string — models are
    /// not always consistent, and a quoted "3" should not lose the count.
    /// </summary>
    private static int? Integer(JsonElement row, string name)
    {
        if (!row.TryGetProperty(name, out var value)) return null;
        return value.ValueKind switch
        {
            JsonValueKind.Number when value.TryGetInt32(out var number) => number,
            JsonValueKind.Number when value.TryGetDouble(out var real) => (int)real,
            JsonValueKind.String when int.TryParse(value.GetString(), out var parsed) => parsed,
            _ => null,
        };
    }

    /// <summary>The human-readable half of an OpenAI error body, when there is one.</summary>
    internal static string? MessageIn(string data)
    {
        try
        {
            using var document = JsonDocument.Parse(data);
            if (document.RootElement.ValueKind == JsonValueKind.Object
                && document.RootElement.TryGetProperty("error", out var error)
                && error.TryGetProperty("message", out var message)
                && message.ValueKind == JsonValueKind.String)
            {
                return message.GetString();
            }
        }
        catch (JsonException)
        {
            // A non-JSON error body (a proxy's HTML, say) has no message to
            // show; the status code stands on its own.
        }
        return null;
    }
}

/// <summary>
/// A stand-in interpreter for tests: answers instantly from the built-in
/// parser instead of reaching the network, so nothing that renders the UI can
/// make a request or need a key.
/// </summary>
public sealed class PreviewExerciseInterpreter : IExerciseInterpreter
{
    private readonly IReadOnlyList<Exercise>? _result;

    /// <param name="result">What to return; <c>null</c> parses the text locally.</param>
    public PreviewExerciseInterpreter(IReadOnlyList<Exercise>? result = null) => _result = result;

    public Task<IReadOnlyList<Exercise>> InterpretAsync(
        string text, CancellationToken cancellationToken = default) =>
        Task.FromResult(_result ?? ExerciseImporter.Parse(text));
}
